util/mutex.go

```diff
@@ -0,0 +1,108 @@
+package util
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+// RWLocker defines the interface for a read-write lock like sync.RWMutex
+type RWLocker interface {
+	sync.Locker
+	RLock()
+	TryRLock() bool
+	RUnlock()
+	TryLock() bool
+	RLocker() sync.Locker
+}
+
+var _ RWLocker = &sync.RWMutex{}
+
+// ContextRWLocker returns a RWMutex that supports context cancellation.
+type ContextRWLocker struct {
+	mux sync.RWMutex
+}
+
+// LockContext locks the RWLocker with the given context.
+// If the context is canceled the function will return immediately with the error from the context.
+// It will still acquire the lock on the background, but it will be released immediately.
+func (c *ContextRWLocker) LockContext(ctx context.Context) error {
+	return lockWithCancel(ctx, c.mux.Lock, c.mux.Unlock)
+}
+
+// RLockContext locks the RWLocker with the given context.
+// If the context is canceled the function will return immediately with the error from the context.
+// It will still acquire the lock on the background, but it will be released immediately.
+func (c *ContextRWLocker) RLockContext(ctx context.Context) error {
+	return lockWithCancel(ctx, c.mux.RLock, c.mux.RUnlock)
+}
+
+func lockWithCancel(ctx context.Context, fnLock func(), fnUnlock func()) error {
+	locked := make(chan bool)
+	expired := &atomic.Value{}
+
+	// We need an additional mutex to synchronize signalling the locking goroutine the context expired.
+	// Otherwise, if the context cancellation handler and locking goroutine execute concurrently,
+	// the context cancellation handler might return an error (which should cause an immediate unlock after acquiring the lock),
+	// which (the signal) might be lost when the locking goroutine is already finished (or just finishing).
+	m := &sync.Mutex{}
+
+	go func() {
+		fnLock()
+		m.Lock()
+		if expired.Load() != nil {
+			// Context expired, unlock immediately.
+			fnUnlock()
+		} else {
+			locked <- true
+		}
+		m.Unlock()
```

Comment:
usually it's a `defer` right after the lock

Reply:
I guess the few extra nanoseconds `defer` adds won't really hurt, given the lock is used for securing expensive database access.